File: android/app/src/main/kotlin/org/recall/mic/SegmentStore.kt
package org.recall.mic

import java.io.File

/**
 * The phone's segment cache: where closed segments wait, and what each
 * directory MEANS — the meeting queue's idiom, because every state here is a
 * verdict that must survive a reboot and a rename is the only change that
 * cannot half-happen:
 *
 * | directory              | meaning                                             |
 * |------------------------|-----------------------------------------------------|
 * | `segments/open/`       | being written by the recorder, nothing touches it   |
 * | `segments/`            | closed, undelivered — what the uploader drains      |
 * | `segments/delivered/`  | Isis holds it, PROVEN: the receipt's sha-256 matched a local re-hash |
 * | `segments/conflict/`   | Isis holds DIFFERENT bytes under this name — a person must look |
 *
 * Deletion happens in exactly one place, [evict], and eats only `delivered/`,
 * oldest first, under cache pressure — never an undelivered segment, never a
 * conflict, never because a server asked (recall/docs/architecture.md,
 * decision 2: eviction is a local decision; Isis's word destroys nothing).
 */
object SegmentStore {
    private const val DIR = "segments"
    private const val OPEN = "open"
    private const val DELIVERED = "delivered"
    private const val CONFLICT = "conflict"

    /** ~2 GiB: hours of WAV, days of FLAC — enough local history to span the
     * upload→nightly-backup window that makes eviction safe at all. */
    const val DEFAULT_CEILING_BYTES = 2L * 1024 * 1024 * 1024

    fun root(base: File): File = File(base, DIR).apply { mkdirs() }

    fun open(base: File): File = File(root(base), OPEN).apply { mkdirs() }

    fun delivered(base: File): File = File(root(base), DELIVERED).apply { mkdirs() }

    fun conflict(base: File): File = File(root(base), CONFLICT).apply { mkdirs() }

    private fun segmentsIn(dir: File): List<File> =
        (dir.listFiles() ?: emptyArray()).filter { it.isFile }.sortedBy { it.name }

    /** Closed, undelivered, oldest first — the uploader's work list. */
    fun undelivered(base: File): List<File> = segmentsIn(root(base))

    /**
     * Adopt anything a crash left in `open/`: a truncated segment is real
     * audio that was really heard — completeness outranks tidiness, so it
     * closes as-is and ships. Run once at recorder start, before a new
     * segment opens.
     */
    fun sweepOpen(base: File) {
        for (orphan in segmentsIn(open(base))) {
            orphan.renameTo(File(root(base), orphan.name))
        }
    }

    fun markDelivered(base: File, segment: File): Boolean =
        segment.renameTo(File(delivered(base), segment.name))

    fun markConflict(base: File, segment: File): Boolean =
        segment.renameTo(File(conflict(base), segment.name))

    private fun totalBytes(base: File): Long =
        (
            segmentsIn(root(base)) + segmentsIn(open(base)) +
                segmentsIn(delivered(base)) + segmentsIn(conflict(base))
        ).sumOf { it.length() }

    /**
     * Free space down to [ceilingBytes], eating verified-delivered segments
     * oldest first and NOTHING else. Returns how many were evicted. If the
     * cache is over the ceiling with `delivered/` empty, it stays over — the
     * undelivered audio is the point of the cache, and deleting it to meet a
     * number would be the exact loss this design exists to prevent.
     */
    fun evict(base: File, ceilingBytes: Long = DEFAULT_CEILING_BYTES): Int {
        var evicted = 0
        var total = totalBytes(base)
        for (oldest in segmentsIn(delivered(base))) {
            if (total <= ceilingBytes) break
            val size = oldest.length()
            if (oldest.delete()) {
                total -= size
                evicted += 1
            }
        }
        return evicted
    }
}

File: ios/Sources/SegmentStore.swift
import Foundation

/// The phone's segment cache (docs/architecture.md, stage C2) — the meeting
/// queue's idiom shared with Android: a segment's state is which directory it
/// is in, because a rename cannot half-happen.
///
///   segments/open/       being written; nothing touches it
///   segments/            closed, undelivered — what the uploader drains
///   segments/delivered/  Isis holds it, PROVEN: receipt sha-256 matched a local re-hash
///   segments/conflict/   Isis holds DIFFERENT bytes under this name — a person must look
///
/// Deletion happens only in `evict`, eats only `delivered/`, oldest first,
/// under cache pressure — never an undelivered segment, never on a server's
/// word (decision 2: eviction is a local decision).
enum SegmentStore {
    /// ~2 GiB: hours of WAV — local history spanning the upload→backup window.
    static let ceilingBytes: Int64 = 2 * 1024 * 1024 * 1024

    static func root() -> URL {
        let base = FileManager.default.urls(
            for: .applicationSupportDirectory, in: .userDomainMask)[0]
        let dir = base.appendingPathComponent("segments", isDirectory: true)
        try? FileManager.default.createDirectory(at: dir, withIntermediateDirectories: true)
        return dir
    }

    static func open() -> URL { sub("open") }
    static func delivered() -> URL { sub("delivered") }
    static func conflict() -> URL { sub("conflict") }

    private static func sub(_ name: String) -> URL {
        let dir = root().appendingPathComponent(name, isDirectory: true)
        try? FileManager.default.createDirectory(at: dir, withIntermediateDirectories: true)
        return dir
    }

    private static func files(in dir: URL) -> [URL] {
        let all =
            (try? FileManager.default.contentsOfDirectory(
                at: dir, includingPropertiesForKeys: [.fileSizeKey])) ?? []
        return all.filter { !$0.hasDirectoryPath }.sorted {
            $0.lastPathComponent < $1.lastPathComponent
        }
    }

    /// Closed, undelivered, oldest first — the uploader's work list.
    static func undelivered() -> [URL] { files(in: root()) }

    /// Adopt a crash's leftovers: a truncated segment is real audio that was
    /// really heard. Run at recorder start, before a new segment opens.
    static func sweepOpen() {
        for orphan in files(in: open()) {
            try? FileManager.default.moveItem(
                at: orphan, to: root().appendingPathComponent(orphan.lastPathComponent))
        }
    }

    static func markDelivered(_ segment: URL) {
        try? FileManager.default.moveItem(
            at: segment, to: delivered().appendingPathComponent(segment.lastPathComponent))
    }

    static func markConflict(_ segment: URL) {
        try? FileManager.default.moveItem(
            at: segment, to: conflict().appendingPathComponent(segment.lastPathComponent))
    }

    private static func size(_ url: URL) -> Int64 {
        Int64((try? url.resourceValues(forKeys: [.fileSizeKey]).fileSize) ?? 0)
    }

    /// Free space down to the ceiling, verified-delivered oldest first and
    /// nothing else. Over the ceiling with nothing delivered stays over —
    /// undelivered audio is the point of the cache.
    static func evict(ceiling: Int64 = ceilingBytes) {
        var total = [root(), open(), delivered(), conflict()]
            .flatMap(files(in:)).reduce(Int64(0)) { $0 + size($1) }
        for oldest in files(in: delivered()) where total > ceiling {
            let bytes = size(oldest)
            if (try? FileManager.default.removeItem(at: oldest)) != nil {
                total -= bytes
            }
        }
    }
}
